File: nexus-mcp/src/nexus_mcp/tools/notes.py
from typing import Any

from nexus_mcp.context import NexusContext
from nexus_mcp.internal_api import internal_api
from nexus_mcp.tools.common import ToolSpec, clean_params, clean_payload, preview

NOTE_FIELDS = {
    "title",
    "content",
    "parent_id",
    "template_id",
    "tags",
    "cover_image",
    "icon",
    "is_public",
    "attachments",
}


async def list_notes(args: dict[str, Any], context: NexusContext) -> Any:
    return await internal_api.list_notes(
        context,
        clean_params(args, {"parent_id", "is_deleted", "is_archived"}),
    )


async def search_notes(args: dict[str, Any], context: NexusContext) -> Any:
    query = args.get("q")
    if not isinstance(query, str) or not query:
        raise ValueError("q is required")
    params = clean_params(args, {"q", "mode", "limit", "offset"})
    return await internal_api.search_notes(context, params)


async def create_note(args: dict[str, Any], context: NexusContext) -> Any:
    payload = clean_payload(args, NOTE_FIELDS)
    if not payload.get("title"):
        raise ValueError("title is required")

    if not context.execute_actions:
        return preview("create_note", payload)
    return await internal_api.create_note(context, payload)


async def get_note(args: dict[str, Any], context: NexusContext) -> Any:
    note_id = args.get("note_id")
    if not isinstance(note_id, str) or not note_id:
        raise ValueError("note_id is required")
    return await internal_api.get_note(context, note_id)


async def update_note(args: dict[str, Any], context: NexusContext) -> Any:
    note_id = args.get("note_id")
    if not isinstance(note_id, str) or not note_id:
        raise ValueError("note_id is required")
    payload = clean_payload(args, NOTE_FIELDS)
    if not payload:
        raise ValueError("At least one note field is required")

    if not context.execute_actions:
        return preview("update_note", {"note_id": note_id, **payload})
    return await internal_api.update_note(context, note_id, payload)


async def delete_note(args: dict[str, Any], context: NexusContext) -> Any:
    note_id = args.get("note_id")
    if not isinstance(note_id, str) or not note_id:
        raise ValueError("note_id is required")

    if not context.execute_actions:
        return preview("delete_note", {"note_id": note_id})
    return await internal_api.delete_note(context, note_id)


async def _note_simple_action(args: dict[str, Any], context: NexusContext, action: str) -> Any:
    note_id = args.get("note_id")
    if not isinstance(note_id, str) or not note_id:
        raise ValueError("note_id is required")

    if not context.execute_actions:
        return preview(f"{action}_note", {"note_id": note_id})
    return await internal_api.note_action(context, note_id, action)


async def archive_note(args: dict[str, Any], context: NexusContext) -> Any:
    return await _note_simple_action(args, context, "archive")


async def unarchive_note(args: dict[str, Any], context: NexusContext) -> Any:
    return await _note_simple_action(args, context, "unarchive")


async def restore_note(args: dict[str, Any], context: NexusContext) -> Any:
    return await _note_simple_action(args, context, "restore")


async def duplicate_note(args: dict[str, Any], context: NexusContext) -> Any:
    note_id = args.get("note_id")
    if not isinstance(note_id, str) or not note_id:
        raise ValueError("note_id is required")
    payload = clean_payload(args, {"title", "parent_id", "include_sub_notes"})

    if not context.execute_actions:
        return preview("duplicate_note", {"note_id": note_id, **payload})
    return await internal_api.note_action(context, note_id, "duplicate", payload)


async def share_note(args: dict[str, Any], context: NexusContext) -> Any:
    note_id = args.get("note_id")
    user_ids = args.get("user_ids")
    if not isinstance(note_id, str) or not note_id:
        raise ValueError("note_id is required")
    if not isinstance(user_ids, list) or not user_ids:
        raise ValueError("user_ids is required")

    payload = clean_payload(args, {"user_ids", "permission"})
    if not context.execute_actions:
        return preview("share_note", {"note_id": note_id, **payload})
    return await internal_api.note_action(context, note_id, "share", payload)


async def merge_notes(args: dict[str, Any], context: NexusContext) -> Any:
    note_ids = args.get("note_ids")
    title = args.get("title")
    if not isinstance(note_ids, list) or len(note_ids) < 2:
        raise ValueError("note_ids must include at least two notes")
    if not isinstance(title, str) or not title:
        raise ValueError("title is required")

    payload = clean_payload(
        args,
        {"note_ids", "title", "include_headers", "add_dividers", "sort_by_date", "parent_id"},
    )
    if not context.execute_actions:
        return preview("merge_notes", payload)
    return await internal_api.merge_notes(context, payload)


NOTE_PROPERTIES = {
    "title": {"type": "string"},
    "content": {"type": "string"},
    "parent_id": {"type": "string"},
    "template_id": {"type": "string"},
    "tags": {"type": "array", "items": {"type": "string"}},
    "cover_image": {"type": "string"},
    "icon": {"type": "string"},
    "is_public": {"type": "boolean"},
}

NOTE_ID_SCHEMA = {
    "type": "object",
    "required": ["note_id"],
    "properties": {"note_id": {"type": "string"}},
}


TOOLS = [
    ToolSpec(
        name="list_notes",
        description="List notes in the current workspace with optional parent/deleted/archive filters.",
        input_schema={
            "type": "object",
            "properties": {
                "parent_id": {"type": "string"},
                "is_deleted": {"type": "boolean"},
                "is_archived": {"type": "boolean"},
            },
        },
        handler=list_notes,
    ),
    ToolSpec(
        name="search_notes",
        description="Search notes using keyword, semantic, or hybrid mode.",
        input_schema={
            "type": "object",
            "required": ["q"],
            "properties": {
                "q": {"type": "string"},
                "mode": {"type": "string", "enum": ["keyword", "semantic", "hybrid"]},
                "limit": {"type": "integer"},
                "offset": {"type": "integer"},
            },
        },
        handler=search_notes,
    ),
    ToolSpec(
        name="create_note",
        description="Create a note. Returns preview unless executeActions=true.",
        input_schema={
            "type": "object",
            "required": ["title"],
            "properties": NOTE_PROPERTIES,
        },
        handler=create_note,
    ),
    ToolSpec(
        name="get_note",
        description="Get note details.",
        input_schema=NOTE_ID_SCHEMA,
        handler=get_note,
    ),
    ToolSpec(
        name="update_note",
        description="Update note fields. Returns preview unless executeActions=true.",
        input_schema={
            "type": "object",
            "required": ["note_id"],
            "properties": {"note_id": {"type": "string"}, **NOTE_PROPERTIES},
        },
        handler=update_note,
    ),
    ToolSpec(
        name="delete_note",
        description="Soft-delete a note. Returns preview unless executeActions=true.",
        input_schema=NOTE_ID_SCHEMA,
        handler=delete_note,
    ),
    ToolSpec(
        name="archive_note",
        description="Archive a note. Returns preview unless executeActions=true.",
        input_schema=NOTE_ID_SCHEMA,
        handler=archive_note,
    ),
    ToolSpec(
        name="unarchive_note",
        description="Unarchive a note. Returns preview unless executeActions=true.",
        input_schema=NOTE_ID_SCHEMA,
        handler=unarchive_note,
    ),
    ToolSpec(
        name="restore_note",
        description="Restore a soft-deleted note. Returns preview unless executeActions=true.",
        input_schema=NOTE_ID_SCHEMA,
        handler=restore_note,
    ),
    ToolSpec(
        name="duplicate_note",
        description="Duplicate a note. Returns preview unless executeActions=true.",
        input_schema={
            "type": "object",
            "required": ["note_id"],
            "properties": {
                "note_id": {"type": "string"},
                "title": {"type": "string"},
                "parent_id": {"type": "string"},
                "include_sub_notes": {"type": "boolean"},
            },
        },
        handler=duplicate_note,
    ),
    ToolSpec(
        name="share_note",
        description="Share a note with workspace users. Returns preview unless executeActions=true.",
        input_schema={
            "type": "object",
            "required": ["note_id", "user_ids"],
            "properties": {
                "note_id": {"type": "string"},
                "user_ids": {"type": "array", "items": {"type": "string"}},
                "permission": {"type": "string", "enum": ["read", "write", "admin"]},
            },
        },
        handler=share_note,
    ),
    ToolSpec(
        name="merge_notes",
        description="Merge multiple notes into a new note. Returns preview unless executeActions=true.",
        input_schema={
            "type": "object",
            "required": ["note_ids", "title"],
            "properties": {
                "note_ids": {"type": "array", "items": {"type": "string"}},
                "title": {"type": "string"},
                "include_headers": {"type": "boolean"},
                "add_dividers": {"type": "boolean"},
                "sort_by_date": {"type": "boolean"},
                "parent_id": {"type": "string"},
            },
        },
        handler=merge_notes,
    ),
]

File: nexus-mcp/src/tools/notes.ts
import type { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { z } from 'zod';
import { idSchema, limitSchema, offsetSchema, workspaceIdSchema } from '../schemas/common.js';
import {
  createNoteInputShape,
  createNoteOutputSchema,
  noteArchiveOutputSchema,
  noteDeleteOutputSchema,
  noteOutputSchema,
  noteRestoreOutputSchema,
  notesListOutputSchema,
  shareNoteInputShape,
  shareNoteOutputSchema,
  searchNotesListOutputSchema,
  updateNoteInputShape,
  noteUnarchiveOutputSchema,
} from '../schemas/notes.js';
import type { NexusApiClient } from '../services/nexus-api.js';
import { registerApiTool } from './register-api-tool.js';

const readOnly = {
  readOnlyHint: true,
  destructiveHint: false,
  idempotentHint: true,
  openWorldHint: true,
};

const write = {
  readOnlyHint: false,
  destructiveHint: false,
  idempotentHint: false,
  openWorldHint: true,
};

export function registerNotesTools(server: McpServer, client: NexusApiClient) {
  registerApiTool(server, client, {
    name: 'nexus_list_notes',
    title: 'List Nexus Notes',
    description:
      'Use this tool when you need to list workspace notes in the Nexus Notes module. These are shared workspace documents, not private AI memory. Filter by parent_id for hierarchical browsing, is_deleted to include soft-deleted notes, or is_archived to show/hide archived notes.',
    inputSchema: {
      workspace_id: workspaceIdSchema,
      parent_id: z.string().optional().describe('Optional parent note/folder ID.'),
      is_deleted: z.boolean().optional().describe('True for deleted notes, false for active notes.'),
      is_archived: z.boolean().optional().describe('True for archived notes, false for non-archived notes.'),
    },
    path: ({ workspace_id }) => `workspaces/${encodeURIComponent(String(workspace_id))}/notes`,
    query: ({ parent_id, is_deleted, is_archived }) => ({ parent_id, is_deleted, is_archived }),
    outputSchema: notesListOutputSchema,
    outputTransform: (data) => ({ notes: data }),
    annotations: readOnly,
  });

  registerApiTool(server, client, {
    name: 'nexus_search_notes',
    title: 'Search Nexus Notes',
    description:
      'Use this tool when you need to search workspace notes in the Nexus Notes module using keyword, semantic (AI embedding), or hybrid mode. These results are workspace documents, not private AI memory. Provide a search query and optionally paginate with limit/offset.',
    inputSchema: {
      workspace_id: workspaceIdSchema,
      q: z.string().min(1).describe('Search query text.'),
      mode: z.enum(['keyword', 'semantic', 'hybrid']).default('hybrid').describe('Search mode.'),
      limit: limitSchema,
      offset: offsetSchema,
    },
    path: ({ workspace_id }) => `workspaces/${encodeURIComponent(String(workspace_id))}/notes/search`,
    query: ({ q, mode, limit, offset }) => ({ q, mode, limit, offset }),
    outputSchema: searchNotesListOutputSchema,
    outputTransform: (data) => ({ notes: data }),
    annotations: readOnly,
  });

  registerApiTool(server, client, {
    name: 'nexus_get_note',
    title: 'Get Nexus Note',
    description:
      'Use this tool when you need to get a workspace note by ID from the Nexus Notes module. Returns full note content, metadata, author info, collaborators, and attachments. This is for persisted workspace documents, not private AI memory. Optionally include soft-deleted notes with include_deleted=true.',
    inputSchema: {
      workspace_id: workspaceIdSchema,
      note_id: idSchema('Note'),
      include_deleted: z.boolean().default(false).describe('Whether to include soft-deleted notes.'),
    },
    path: ({ workspace_id, note_id }) =>
      `workspaces/${encodeURIComponent(String(workspace_id))}/notes/${encodeURIComponent(String(note_id))}`,
    query: ({ include_deleted }) => ({ include_deleted }),
    outputSchema: noteOutputSchema,
    annotations: readOnly,
  });

  registerApiTool(server, client, {
    name: 'nexus_create_note',
    title: 'Create Nexus Note',
    description:
      'Use this tool when you need to create a new workspace note in the Nexus Notes module. Use it for meeting notes, shared documentation, or other workspace-visible documents. Do not use it to store private AI memory, user preferences, or future-chat reminders. Provide title and content (required), and optionally set a parent_id for nesting, tags, cover_image, icon, is_public flag, and file/note/event attachments.',
    inputSchema: { workspace_id: workspaceIdSchema, ...createNoteInputShape },
    method: 'POST',
    path: ({ workspace_id }) => `workspaces/${encodeURIComponent(String(workspace_id))}/notes`,
    body: ({ title, content, parent_id, tags, cover_image, icon, is_public, attachments }) => ({
      title,
      content,
      ...(parent_id !== undefined ? { parent_id } : {}),
      ...(tags !== undefined ? { tags } : {}),
      ...(cover_image !== undefined ? { cover_image } : {}),
      ...(icon !== undefined ? { icon } : {}),
      ...(is_public !== undefined ? { is_public } : {}),
      ...(attachments !== undefined ? { attachments } : {}),
    }),
    outputSchema: createNoteOutputSchema,
    annotations: write,
  });

  registerApiTool(server, client, {
    name: 'nexus_update_note',
    title: 'Update Nexus Note',
    description:
      'Use this tool when you need to update an existing workspace note in the Nexus Notes module. Only use it when the user wants to change a persisted workspace document. Do not use it for private AI memory, user preferences, or future-chat reminders. All fields are optional—only provided fields will be updated.',
    inputSchema: { workspace_id: workspaceIdSchema, note_id: idSchema('Note'), ...updateNoteInputShape },
    method: 'PATCH',
    path: ({ workspace_id, note_id }) =>
      `workspaces/${encodeURIComponent(String(workspace_id))}/notes/${encodeURIComponent(String(note_id))}`,
    body: ({ title, content, tags, is_public, is_favorite, attachments }) => ({
      ...(title !== undefined ? { title } : {}),
      ...(content !== undefined ? { content } : {}),
      ...(tags !== undefined ? { tags } : {}),
      ...(is_public !== undefined ? { is_public } : {}),
      ...(is_favorite !== undefined ? { is_favorite } : {}),
      ...(attachments !== undefined ? { attachments } : {}),
    }),
    outputSchema: createNoteOutputSchema,
    annotations: write,
  });

  registerApiTool(server, client, {
    name: 'nexus_archive_note',
    title: 'Archive Nexus Note',
    description:
      'Use this tool when you need to archive a note and all its sub-notes to hide them from the default list view. Archived notes can be unarchived later.',
    inputSchema: { workspace_id: workspaceIdSchema, note_id: idSchema('Note') },
    method: 'POST',
    path: ({ workspace_id, note_id }) =>
      `workspaces/${encodeURIComponent(String(workspace_id))}/notes/${encodeURIComponent(String(note_id))}/archive`,
    outputSchema: noteArchiveOutputSchema,
    annotations: {
      readOnlyHint: false,
      destructiveHint: false,
      idempotentHint: false,
      openWorldHint: true,
    },
  });

  registerApiTool(server, client, {
    name: 'nexus_delete_note',
    title: 'Delete Nexus Note',
    description:
      'Use this tool when you need to soft delete a note and all its sub-notes. Deleted notes can be restored or viewed with include_deleted=true on list/get queries.',
    inputSchema: { workspace_id: workspaceIdSchema, note_id: idSchema('Note') },
    method: 'DELETE',
    path: ({ workspace_id, note_id }) =>
      `workspaces/${encodeURIComponent(String(workspace_id))}/notes/${encodeURIComponent(String(note_id))}`,
    outputSchema: noteDeleteOutputSchema,
    annotations: {
      readOnlyHint: false,
      destructiveHint: false,
      idempotentHint: false,
      openWorldHint: true,
    },
  });

  registerApiTool(server, client, {
    name: 'nexus_restore_note',
    title: 'Restore Nexus Note',
    description:
      'Use this tool when you need to restore a soft-deleted note and all its sub-notes back to active status.',
    inputSchema: { workspace_id: workspaceIdSchema, note_id: idSchema('Note') },
    method: 'POST',
    path: ({ workspace_id, note_id }) =>
      `workspaces/${encodeURIComponent(String(workspace_id))}/notes/${encodeURIComponent(String(note_id))}/restore`,
    outputSchema: noteRestoreOutputSchema,
    annotations: {
      readOnlyHint: false,
      destructiveHint: false,
      idempotentHint: false,
      openWorldHint: true,
    },
  });

  registerApiTool(server, client, {
    name: 'nexus_unarchive_note',
    title: 'Unarchive Nexus Note',
    description:
      'Use this tool when you need to unarchive a previously archived note and all its sub-notes, making them visible in the default list view again.',
    inputSchema: { workspace_id: workspaceIdSchema, note_id: idSchema('Note') },
    method: 'POST',
    path: ({ workspace_id, note_id }) =>
      `workspaces/${encodeURIComponent(String(workspace_id))}/notes/${encodeURIComponent(String(note_id))}/unarchive`,
    outputSchema: noteUnarchiveOutputSchema,
    annotations: {
      readOnlyHint: false,
      destructiveHint: false,
      idempotentHint: false,
      openWorldHint: true,
    },
  });

  registerApiTool(server, client, {
    name: 'nexus_share_note',
    title: 'Share Nexus Note',
    description:
      'Use this tool when you need to share a workspace note with workspace members. This applies to Nexus Notes documents, not private AI memory. Provide an array of user_ids and optionally set a permission level (read/write/admin).',
    inputSchema: { workspace_id: workspaceIdSchema, note_id: idSchema('Note'), ...shareNoteInputShape },
    method: 'POST',
    path: ({ workspace_id, note_id }) =>
      `workspaces/${encodeURIComponent(String(workspace_id))}/notes/${encodeURIComponent(String(note_id))}/share`,
    body: ({ user_ids, permission }) => ({
      user_ids,
      ...(permission !== undefined ? { permission } : {}),
    }),
    outputSchema: shareNoteOutputSchema,
    annotations: write,
  });
}
